src/algebra/category/Module/change_of_rings.lean

```diff
@@ -160,4 +164,80 @@ variables {R : Type u₁} {S : Type u₂} [comm_ring R] [comm_ring S] (f : R →
+
+section unbundled
+
+variables (M : Type v) [add_comm_monoid M] [module R M]
+
+local notation `Hom` M := (restrict_scalars f).obj ⟨S⟩ →ₗ[R] M
+
+/--
+ Given an `R`-module M, consider the Hom(S, M) -- the `R`-linear maps between S (as an `R`-module by
+ means of restriction of scalars) and M. `S` acts on Hom(S, M) by `s • g = x ↦ g (x • s)`
+ -/
+instance has_smul : has_smul S $ Hom M :=
+{ smul := λ s g,
+  { to_fun := λ (s' : S), g (s' • s : S),
+    map_add' := λ (x y : S), by simp [add_smul, map_add],
+    map_smul' := λ r (t : S), by rw [ring_hom.id_apply, @restrict_scalars.smul_def _ _ _ _ f ⟨S⟩,
+      ←linear_map.map_smul, @restrict_scalars.smul_def _ _ _ _ f ⟨S⟩, smul_assoc] } }
+
+@[simp] lemma smul_apply (s : S) (g : Hom M) (s' : S) :
+  @has_smul.smul _ _ (coextend_scalars.has_smul f _) s g s' = g (s' • s : S) := rfl
+
+instance mul_action : mul_action S $ Hom M :=
+{ one_smul := λ g, linear_map.ext $ λ (s : S), by simp,
+  mul_smul := λ (s t : S) g, linear_map.ext $ λ (x : S), by simp [mul_assoc],
+  ..coextend_scalars.has_smul f _ }
+
+instance distrib_mul_action : distrib_mul_action S $ Hom M :=
+{ smul_add := λ s g h, linear_map.ext $ λ (t : S), by simp,
+  smul_zero := λ s, linear_map.ext $ λ (t : S), by simp,
+  ..coextend_scalars.mul_action f _ }
+
+/--
+`S` acts on Hom(S, M) by `s • g = x ↦ g (x • s)`, this action defines an `S`-module structure on
+Hom(S, M).
+ -/
+instance is_module : module S $ Hom M :=
+{ add_smul := λ s1 s2 g, linear_map.ext $ λ (x : S), by simp [mul_add],
+  zero_smul := λ g, linear_map.ext $ λ (x : S), by simp,
+  ..coextend_scalars.distrib_mul_action f _ }
+
+end unbundled
+
+variable (M : Module.{v} R)
+
+/-- If `M` is an `R`-module, then the set of `R`-linear maps `S →ₗ[R] M` is an `S`-module with
+scalar multiplication defined by `s • l := x ↦ l (x • s)`-/
+def obj' : Module S := ⟨(restrict_scalars f).obj ⟨S⟩ →ₗ[R] M⟩
+
+instance : has_coe_to_fun (obj' f M) (λ g, S → M) :=
+{ coe := λ g, g.to_fun }
+
+/-- If `M, M'` are `R`-modules, then any `R`-linear map `g : M ⟶ M'` induces an `S`-linear map
+`(S →ₗ[R] M) ⟶ (S →ₗ[R] M')` defined by `h ↦ g ∘ h`-/
+@[simps] def map' {M M' : Module R} (g : M ⟶ M') : obj' f M ⟶ obj' f M' :=
+{ to_fun := λ h, g.comp h,
+  map_add' := λ _ _, linear_map.comp_add _ _ _,
+  map_smul' := λ s h, linear_map.ext $ λ (t : S), by simpa only [smul_apply] }
+
+end coextend_scalars
+
+/--
+For any rings `R, S` and a ring homomorphism `f : R →+* S`, there is a functor from `R`-module to
+`S`-module defined by `M ↦ (S →ₗ[R] M)` where `S` is considered as an `R`-module via restriction of
+scalar and `g : M ⟶ M'` is sent to `h ↦ g ∘ h`.
+-/
+def coextend_scalars {R : Type u₁} {S : Type u₂} [ring R] [ring S] (f : R →+* S) :
+  Module R ⥤ Module S :=
+{ obj := coextend_scalars.obj' f,
+  map := λ _ _, coextend_scalars.map' f,
+  map_id' := λ M, linear_map.ext $ λ h, linear_map.ext $ λ x, rfl,
+  map_comp' := λ _ _ _ g h, linear_map.ext $ λ h, linear_map.ext $ λ x, rfl }
+
```

Comment:
Do we want to give the coercion to fun here? Something like
```suggestion
section
variables {R : Type u₁} {S : Type u₂} [ring R] [ring S] (f : R →+* S)

instance (M : Module R) : has_coe_to_fun ((coextend_scalars f).obj M) (λ g, S → M) :=
(infer_instance : has_coe_to_fun (coextend_scalars.obj' f M) _)

@[simp] lemma smul_apply (M : Module R) (m : (coextend_scalars f).obj M) (s s' : S) :
  (s • m) s' = m (s' * s) := rfl

end
```

Reply:
Yes, this is a good idea!